src/agent_writer.cpp

```diff
@@ -156,12 +156,35 @@ void AgentWriter::startWriting(std::unique_ptr<Handle> handle) {
             payload = trace_encoder_->payload();
             trace_encoder_->clearTraces();
           }  // lock on mutex_ ends.
-          // Send spans, not in critical period.
+          // Send the spans. Note that this is not in the critical section above.
           bool success = retryFiniteOnFail(
               [&]() { return AgentWriter::postTraces(handle, headers, payload, logger_); });
+          // Sending could fail. If it succeeds, then the HTTP response status
+          // could indicate an error or success.
           if (success) {
-            trace_encoder_->handleResponse(handle->getResponse());
+            const int response_status = handle->getResponseStatus();
+            if (response_status == 0) {
+              const std::string body = handle->getResponse();
+              std::ostringstream diagnostic;
+              diagnostic << "Datadog Agent returned response without an HTTP status and with the "
+                            "following body of length "
+                         << body.size() << ": " << body;
+              logger_->Log(LogLevel::error, diagnostic.str());
+            } else if (response_status < 200 || response_status >= 300) {
```

Comment:
Might want to be more specific about valid statuses here. Pretty sure the agent would only normally send a 200, and anything else would be surprising.

Reply:
The source code agrees with you: https://github.com/DataDog/datadog-agent/blob/main/pkg/trace/api/api.go

However, the status is not part of the documented contract between the agent and tracer libraries (what little contract there is: I did some documentation [here][1]).

I don't know how likely the de facto "only 200" is to change (probably 0%), but let's accept all of the 2xx anyway.

The java tracer checks `[200, 300)` as we do here. References:
- https://github.com/DataDog/dd-trace-java/blob/b25b4321c847f4206ab4970c8b6a9ab4f35ee838/dd-trace-core/src/main/java/datadog/trace/common/writer/ddintake/DDEvpProxyApi.java#LL124C6-L124C6
- https://square.github.io/okhttp/4.x/okhttp/okhttp3/-response/is-successful/

I'll leave it.

[1]: https://github.com/DataDog/datadog-agent/blob/main/pkg/trace/api/version.go

Reply:
The plot thickens: https://github.com/DataDog/datadog-agent/blob/3723c72a732fb26510e4813690d2cc92e2732f2a/pkg/trace/api/api.go#L89-L92